src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -3647,13 +3647,13 @@ public async Task TestFrozenPartialSemanticsHandlesRemoveAndAddWithNullPathAndDi
             project = project.RemoveDocument(project.DocumentIds.Single())
                 .AddDocument("RegularDocument2.cs", "// Source File", filePath: null).Project;
 
-            // Freeze semantics -- with the new document; this should still give us a project with two documents: the new
-            // one will be added, and the old one will stay around since the name differed.
+            // Freeze semantics -- with the new document; this should still give us a project with one document: just
+            // the newly added one.
             var frozenDocument = project.Documents.Single().WithFrozenPartialSemantics(CancellationToken.None);
 
-            Assert.Equal(2, frozenDocument.Project.Documents.Count());
+            Assert.Equal(1, frozenDocument.Project.Documents.Count());
             var treesInCompilation = (await frozenDocument.Project.GetCompilationAsync()).SyntaxTrees;
-            Assert.Equal(2, treesInCompilation.Count());
+            Assert.Equal(1, treesInCompilation.Count());
```

Comment:
put into: https://github.com/dotnet/roslyn/pull/72111

Reply:
@CyrusNajmabadi I was at first confused why this behavior changed but I get it: we were previously comparing the paths of the _trees_ which aren't allowed to be null from the compiler's view of the world, and are filled in with the document name in that case. In the case we have document names but no file paths, it means we're really picking a completely arbitrary document to count as a replacement here. Put another way: if we know both docs have no file paths but different names, why are we considering them the same if the only hint we have says they should be different?

This is still an edge case though, and my original commit comment matches my memory of this test:

> The code change is relatively small, but while I'm here I also tried to make this a bit more robust if a tree didn't have a file path, which was suspiciously forgotten about in the original code.

i.e. we didn't have a specific requirement for what this function should do in this case, but I was just ensuring we weren't crashing or doing something 100% nonsensical. So I'd argue the previous behavior is more correct, but this is a niche scenario -- especially because in this case we have to have a workspace that has no file paths for documents, but is also using partial semantics which we don't have today.

Reply:
I guess at this point we could also just remove the path checking entirely and say hey, it's got a new ID, good enough. I don't know why we originally had the remove-and-add logic. Since the code originally operated on looking at the compilation's syntax trees directly, I don't recall whether this was a scenario we cared about or simply we wanted to ensure we didn't add two trees with the same path by accident.

Reply:
works for me.

Reply:
handled in followup pr.